cpp/include/raft/core/mdbuffer.cuh

```diff
@@ -0,0 +1,831 @@
+   * storage
+   */
+  [[nodiscard]] auto constexpr is_owning() const
+  {
+    return data_.index() >= std::variant_size_v<view_type_variant>;
+  };
+
+ private:
+  template <typename MemTypeConstant>
+  [[nodiscard]] auto view()
+  {
+    if constexpr (MemTypeConstant::value.has_value()) {
+      if (is_owning()) {
+        return std::get<owning_type<MemTypeConstant::value.value()>>(data_).view();
+      } else {
+        return std::get<view_type<MemTypeConstant::value.value()>>(data_);
+      }
+    } else {
+      return std::visit(
+        [](auto&& inner) {
+          if constexpr (is_mdspan_v<std::remove_reference_t<decltype(inner)>>) {
+            return view_type_variant{inner};
+          } else {
+            return view_type_variant{inner.view()};
+          }
+        },
+        data_);
+    }
+  }
+
+  template <typename MemTypeConstant>
+  [[nodiscard]] auto view() const
+  {
+    if constexpr (MemTypeConstant::value.has_value()) {
+      if (is_owning()) {
+        return make_const_mdspan(
+          std::get<owning_type<MemTypeConstant::value.value()>>(data_).view());
+      } else {
+        return make_const_mdspan(std::get<view_type<MemTypeConstant::value.value()>>(data_));
+      }
+    } else {
+      return std::visit(
+        [](auto&& inner) {
+          if constexpr (is_mdspan_v<std::remove_reference_t<decltype(inner)>>) {
+            return const_view_type_variant{make_const_mdspan(inner)};
+          } else {
+            return const_view_type_variant{make_const_mdspan(inner.view())};
+          }
+        },
+        data_);
+    }
+  }
+
+ public:
+  /**
+   * @brief Return an mdspan of the indicated memory type representing a view
+   * on the stored data. If the mdbuffer does not contain data of the indicated
+   * memory type, a std::bad_variant_access will be thrown.
+   */
+  template <memory_type mem_type>
+  [[nodiscard]] auto view()
+  {
+    return view<memory_type_constant<mem_type>>();
+  }
+  /**
+   * @brief Return an mdspan containing const elementgs of the indicated memory type representing a
+   * view on the stored data. If the mdbuffer does not contain data of the indicated memory type, a
+   * std::bad_variant_access will be thrown.
+   */
+  template <memory_type mem_type>
+  [[nodiscard]] auto view() const
+  {
+    return view<memory_type_constant<mem_type>>();
+  }
+  /**
+   * @brief Return a std::variant representing the possible mdspan types that
+   * could be returned as views on the mdbuffer. The variant will contain the mdspan
+   * corresponding to its current memory type.
+   *
+   * This method is useful for writing generic code to handle any memory type
+   * that might be contained in an mdbuffer at a particular point in a
+   * workflow. By performing a `std::visit` on the returned value, the caller
+   * can easily dispatch to the correct code path for the memory type.
+   */
+  [[nodiscard]] auto view() { return view<memory_type_constant<>>(); }
```

Comment:
What does `memory_type_constant<>` return when template parameter is not specified?

Reply:
`std::nullopt` (of type `std::optional<memory_type>`)